File: src/com/haskforce/highlighting/annotation/external/HaskellExternalAnnotator.java
package com.haskforce.highlighting.annotation.external;

import com.haskforce.codeInsight.HaskellCompletionContributor;
import com.haskforce.highlighting.annotation.HaskellAnnotationHolder;
import com.haskforce.highlighting.annotation.HaskellProblem;
import com.haskforce.highlighting.annotation.Problems;
import com.haskforce.settings.ToolKey;
import com.haskforce.utils.ExecUtil;
import com.intellij.lang.annotation.AnnotationHolder;
import com.intellij.lang.annotation.ExternalAnnotator;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ModalityState;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.fileEditor.FileDocumentManager;
import com.intellij.openapi.module.Module;
import com.intellij.openapi.module.ModuleUtilCore;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiFile;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.Future;

/**
 * Single annotator that calls all external tools used for annotations.
 */
public class HaskellExternalAnnotator extends ExternalAnnotator<PsiFile, HaskellExternalAnnotator.State> {
    @SuppressWarnings("UnusedDeclaration")
    private static final Logger LOG = Logger.getInstance(HaskellExternalAnnotator.class);

    /**
     * The default implementation here is to not annotate files that have lexer/parser errors.  This is kind
     * of lame since the error may be invalid.
     */
    @Nullable @Override
    public PsiFile collectInformation(@NotNull PsiFile file, @NotNull Editor editor, boolean hasErrors) {
        return collectInformation(file);
    }

    @NotNull
    @Override
    public PsiFile collectInformation(@NotNull PsiFile file) {
        return file;
    }

    @Nullable
    @Override
    public State doAnnotate(@NotNull PsiFile file) {
        final String canonicalPath = file.getVirtualFile().getCanonicalPath();
        if (canonicalPath == null) {
            return null;
        }

        // Force all files to save to ensure annotations are in sync with the file system.
        ApplicationManager.getApplication().invokeAndWait(new Runnable() {
            @Override
            public void run() {
                FileDocumentManager.getInstance().saveAllDocuments();
            }
        }, ModalityState.any());

        // Use the module directory as working directory if it exists.
        // If not, use the project directory.
        final Project project = file.getProject();
        final String workDir = ExecUtil.guessWorkDir(file);
        HaskellCompletionContributor.loadCacheData(file);
        State state = new State();
        final Module module = ModuleUtilCore.findModuleForPsiElement(file);
        if (module != null) {
            if (ToolKey.GHC_MODI_KEY.getPath(project) != null) {
                GhcModi ghcModi = module.getComponent(GhcModi.class);
                if (ghcModi != null) {
                    state.ghcModiProblems = ghcModi.check(canonicalPath);
                }
            } else {
                state.ghcModProblems = GhcMod.check(module, workDir, canonicalPath);
            }
        }
        state.hlintProblems = HLint.lint(project, workDir, canonicalPath);
        return state;
    }

    /**
     * Wrapper to simplify adding annotations to an AnnotationHolder.
     */
    @Override
    public void apply(@NotNull PsiFile file, State state, @NotNull AnnotationHolder holder) {
        apply(file, state, new HaskellAnnotationHolder(holder));
    }

    public static void apply(@NotNull PsiFile file, State state, @NotNull HaskellAnnotationHolder holder) {
        createAnnotations(file, state.hlintProblems, holder);
        createAnnotations(file, state.ghcModProblems, holder);
        createAnnotations(file, state.ghcModiProblems, holder);
    }

    public static void createAnnotations(@NotNull PsiFile file, @Nullable Problems problems,
                                         @NotNull HaskellAnnotationHolder holder) {
        if (problems == null || problems.isEmpty() || !file.isValid()) return;
        for (HaskellProblem problem : problems) {
            problem.createAnnotations(file, holder);
        }
    }

    public static void createAnnotations(@NotNull PsiFile file, @Nullable Future<Problems> problemsFuture,
                                         @NotNull HaskellAnnotationHolder holder) {
        if (problemsFuture == null) return;
        final Problems problems = GhcModi.getFutureProblems(file.getProject(), problemsFuture);
        if (problems == null) return;
        createAnnotations(file, problems, holder);
    }

    public static class State {
        Problems hlintProblems;
        Problems ghcModProblems;
        Future<Problems> ghcModiProblems;
    }
}

File: src/com/haskforce/haskell/highlighting/annotation/external/HaskellExternalAnnotator.scala
package com.haskforce.highlighting.annotation.external

import scala.collection.JavaConverters._

import com.intellij.lang.annotation.{AnnotationHolder, ExternalAnnotator}
import com.intellij.openapi.application.{ApplicationManager, ModalityState}
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.fileEditor.FileDocumentManager
import com.intellij.psi.PsiFile
import org.jetbrains.annotations.{NotNull, Nullable}

import com.haskforce.Implicits._
import com.haskforce.haskell.codeInsight.HaskellCompletionContributor
import com.haskforce.highlighting.annotation.{HaskellAnnotationHolder, HaskellProblem, Problems}
import HaskellExternalAnnotator.State
import com.haskforce.utils.{SAMUtils, WrappedFuture}

/** Single annotator that calls all external tools used for annotations. */
class HaskellExternalAnnotator extends ExternalAnnotator[PsiFile, State] {

  /**
   * The default implementation here is to not annotate files that have lexer/parser errors.
   * This is kind of lame since the error shouldn't prevent our external annotations from
   * appearing.  To get around this, we ignore the `hasErrors` argument.
   */
  @Nullable
  override def collectInformation
      (@NotNull file: PsiFile, @NotNull editor: Editor, hasErrors: Boolean): PsiFile = {
    collectInformation(file)
  }

  /** Simply returns the file so it will get passed to doAnnotate() */
  @NotNull
  override def collectInformation(@NotNull file: PsiFile): PsiFile = {
    file
  }

  /** Builds the set of annotations to be applied to the source file. */
  @Nullable
  override def doAnnotate(@NotNull file: PsiFile): State = {
    // We need to save the files so external processes will see what we see.
    saveAllFiles()
    // Constructs our annotation state for a file given our registered providers.
    State.buildForFile(file)
  }

  /** Applies the annotations from the State to the source file. */
  override def apply
      (@NotNull file: PsiFile, state: State, @NotNull holder: AnnotationHolder)
      : Unit = {
    HaskellExternalAnnotator.createAnnotations(
      file, state.getProblems, new HaskellAnnotationHolder(holder)
    )
  }

  private def saveAllFiles(): Unit = {
    ApplicationManager.getApplication.invokeAndWait(
      () => FileDocumentManager.getInstance.saveAllDocuments(),
      ModalityState.any()
    )
  }
}

object HaskellExternalAnnotator {

  /** Registered problem providers for our configured tools. */
  val problemsProviderFactories: List[ProblemsProviderFactory] = List(
    LintProblemsProviderFactory,
    CompileProblemsProviderFactory
  )

  /** Helper to annotate problems in our source from external tools. */
  def createAnnotations
      (@NotNull file: PsiFile,
       problems: Stream[HaskellProblem],
       @NotNull holder: HaskellAnnotationHolder)
      : Unit = {
    if (!file.isValid) return
    problems.foreach { _.createAnnotations(file, holder) }
  }

  /** Wraps the problems provided by external tools. */
  final case class State(problems: List[Option[Problems]]) {

    /** Streams the problems in the order they were provided to the State. */
    def getProblems: Stream[HaskellProblem] = {
      problems.toStream.flatten.flatMap(_.iterator().asScala.toStream)
    }
  }

  object State {
    /** Smart constructor which traverses our factories and obtains problem futures. */
    def buildForFile(file: PsiFile): State = {
      State(problemsProviderFactories.flatMap(_.get(file).map(_.getProblems)))
    }
  }
}
